pkg/controller/federatedcluster/clusterclient.go

```diff
@@ -119,48 +120,44 @@ func (self *ClusterClient) GetClusterHealthStatus() *fedv1a1.FederatedClusterSta
 }
 
 // GetClusterZones gets the kubernetes cluster zones and region by inspecting labels on nodes in the cluster.
-func (self *ClusterClient) GetClusterZones() (zone, region string, err error) {
+func (self *ClusterClient) GetClusterZones() ([]string, string, error) {
 	nodes, err := self.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
 	if err != nil {
 		glog.Errorf("Failed to list nodes while getting zone names: %v", err)
-		return "", "", err
+		return nil, "", err
 	}
+
+	zones := sets.NewString()
+	region := ""
 	for i, node := range nodes.Items {
-		zone, err = getZoneNameForNode(node)
-		if err != nil {
-			return "", "", err
-		}
+		zone := getZoneNameForNode(node)
+		// region is same for all nodes in the cluster, so just pick the region from first node.
 		if i == 0 {
-			region, err = getRegionNameForNode(node)
-			if err != nil {
-				return "", "", err
-			}
+			region = getRegionNameForNode(node)
+		}
+		if zone != "" && !zones.Has(zone) {
+			zones.Insert(zone)
 		}
-		// TODO: Optimize this flow. All nodes will have the same zone label.
-		// So just considering first node for now.
-		break
 	}
-	return zone, region, nil
+	return zones.List(), region, nil
 }
 
 // Find the name of the zone in which a Node is running.
-func getZoneNameForNode(node corev1.Node) (string, error) {
+func getZoneNameForNode(node corev1.Node) string {
 	for key, value := range node.Labels {
 		if key == LabelZoneFailureDomain {
-			return value, nil
+			return value
 		}
 	}
-	return "", errors.Errorf("Zone name for node %s not found. No label with key %s",
```

Comment:
(No action required) Why is an error no longer desirable to raise an error if a zone is not found?

Reply:
The error was not much useful here as there could be env where the said lables were not available. I have added reporting similar error in `ServiceDNS` controller.